### ISSUE TITLE
[dart2wasm] Support non-JS wasm runtimes

### ISSUE THREAD

**Issue by eyebrowsoffire** [dart-lang/sdk#53884]
We eventually may want dart2wasm to actually work in non-JavaScript environments (wasmtime, wasmer, etc). @askeksa had mentioned a list of a few things that would definitly need to be addressed in order to remove the dependency on a JS/browser environment:

> We are currently dependent on JS for the following components:
> - Event loop
> - Printing
> - OS-level timer features, such as current time
> - Stack traces
> - Conversions between double and String
> - Regexps
> - Math functions (sin, pow, exp, etc.)
> 
> For each of these, we'll have to either:
> - (1) implement the feature in Dart (or intrinsically in the compiler),
> - (2) link to an implementation compiled to Wasm, or
> - (3) import an external implementation from the host.

**Comment by kripken** [comment]
Can existing JS solutions for those environments help here? That is, [Javy](https://shopify.engineering/javascript-in-webassembly-for-shopify-functions) and [spidermonkey.wasm](https://github.com/bytecodealliance/spidermonkey-wasm-build) and similar projects compile a JS VM into wasm, so it can run in a non-JS wasm environment. If that JS happens to use wasm, it seems like that wasm could be bundled alongside the compiled JS VM, and "just work" with some glue (e.g. spidermonkey already has wasm support; it would just need to be "pointed" to the bundled wasm rather than to interpret it).

The wasm would run at full speed, and the JS is apparently fast enough for those environments, so this seems viable. Perhaps someone has already done it?

If that existed, then dart2wasm and other toolchains would not need to do any special work to port to non-JS environments (no more than people wanting to run JS there, which is considered a solved problem).

**Comment by eyebrowsoffire** [comment]
Yeah I contemplated something like the spidermonkey.wasm might be an interesting idea. Off-hand, it might be even more effective for WasmGC modules with a thin JS support runtime attached, rather than a large blob of JS. I don't know what kind of optimizations are done for the JS VM snapshot build, but a "whole world analysis" on the JS support runtime could probably lead to pruning a *lot* of code out of the JS VM.

**Comment by yjbanov** [comment]
* (4) Exclude the feature, as it may only makes sense within JS environments. For example, the event loop is a very particular type of code execution model that came to most mainstream languages _from_ JS (e.g. Dart, TypeScript, AssemblyScript), which the host will have to support to enable all the event loop type features in Dart (e.g. futures, microtasks, async/await, streams). However, outside the JS environment, there may be sufficiently large categories of programs/libraries that would not need the event loop, and so it may be useful to have a compilation mode that does not rely on it. There's certainly a subset of the Dart language and libraries that don't need it.
* (5) Add the feature as a standard to either Wasm or WASI. Back to the event loop example, if we recognize that it covers a sufficient set of use-cases (such as how it happened for WasmGC), then it may make sense to standardize for event loop-based languages to be able to target all standalone Wasm runtimes that provide it. Dart may not be the only language wanting it. [AssemblyScript seems to want it too](https://github.com/AssemblyScript/assemblyscript/issues/376), and there may be others.

**Comment by askeksa** [comment]
Stack traces could be a good candidate for (5), since it will be difficult to do without support from the engine.

**Comment by zacharypuulsedev** [comment]
This would be amazing to support better serverless, and general flexible compute, using Dart. 

Being able to easily deploy dart applications that are a collection of WASM modules would be revolutionary. Dart is a great language for glueing together libraries quickly, and leveraging existing libraries across languages (esp. rust) that can be compiled to WASM would expand the ecosystem so tremendously. 

While this is already possible with https://pub.dev/packages/wasm_run, it would be great to further the interoperability and deployment process. This would be especially useful when trying to deploy cross platform applications that currently use FFI and native compiled libraries for each supported platform. In the highest performance applications, FFI and native code may still be desirable, but for a vast majority of applications, the ease of quickly composing applications and innovating has huge value.

Advantages 
- Easy interop with other WASM modules/cross language support
  - LLMs/Langchains
  - High performance Rust crates
  - Broad library/language support
- Flexible deployment (Web, Cloud, Native)
- Architecture interoperability (ex. compiling on arm64 and running on x64)
- Ability to run on edge networks
- Faster cold starts (vs docker containers)
- Lower memory (vs docker containers)
- Sandboxing and running sandboxed code

Implementation:
1. SpiderMonkey.wasm (mentioned above seems like a great start)
2. Use rust crates to implement missing functionality; Either compile with binary, or link via WASM modules if the wasm module boundary doesn't have too high of a performance penalty.
    - Event Loop -> [tokio](https://crates.io/crates/tokio)
    - Weak Support -> [weak-table](https://crates.io/crates/weak-table)
    - double to string -> rust stdlib
    - RegEx -> [regex](https://crates.io/crates/regex)
    - math -> rust stdlib or https://lib.rs/science/math
3. Upstream support into WASM/WASI spec for managed languages (Go, Python, Typescript, etc..)
  
--

Regardless of approach taken, facilitating the deployment of applications that are collections of Wasm modules seems like the natural evolution of docker containerization, with even greater platform flexibility, and a higher degree of composability into ever larger applications or pipelines (eg. langchains, llms, sandboxed user code, etc)

**Comment by bivens-dev** [comment]
I saw some mention of SpiderMonkey from @yjbanov and others. I was wondering if this offshoot helps move things any closer as a possibility at all.

> StarlingMonkey is a [SpiderMonkey](https://spidermonkey.dev/) based JS runtime optimized for use in [WebAssembly Components](https://component-model.bytecodealliance.org/). StarlingMonkey's core builtins target WASI 0.2.0 to support a Component Model based event loop and standards-compliant implementations of key web builtins, including the fetch API, WHATWG Streams, text encoding, and others. To support tailoring for specific use cases, it's designed to be highly modular, and can be readily extended with custom builtins and host APIs.

**source**: https://github.com/bytecodealliance/StarlingMonkey

**Comment by iapicca** [comment]
> I saw some mention of SpiderMonkey from @yjbanov and others. I was wondering if this offshoot helps move things any closer as a possibility at all.
> 
> > StarlingMonkey is a [SpiderMonkey](https://spidermonkey.dev/) based JS runtime optimized for use in [WebAssembly Components](https://component-model.bytecodealliance.org/). StarlingMonkey's core builtins target WASI 0.2.0 to support a Component Model based event loop and standards-compliant implementations of key web builtins, including the fetch API, WHATWG Streams, text encoding, and others. To support tailoring for specific use cases, it's designed to be highly modular, and can be readily extended with custom builtins and host APIs.
> 
> **source**: https://github.com/bytecodealliance/StarlingMonkey

is there a way to use dart2JS to compile a non-web target?
it would be very useful to use dart in fermyon

**Comment by maks** [comment]
Yes 💯 @iapicca I have just run into an almost identical use case myself. For my product, I need to compile Dart to JS to run in a embedded JS engine, initially I'm targeting Duktape. I was actually trying to use the DDC for this but the removal of the ability to use it directly via the dart cli command in favour of webserve is *very* annoying and the best I've found so far is to use --no-release with the build command.

I'm very open to even some pointers from the Dart team about where to go look for underlying Dart compiler APIs thar could used as ideally I want to be able to just compile sets of stand alone functions from one or several Dart libraries (ie. files) and not use a main() entry point at all.

**Comment by sigmundch** [comment]
Yes, Dart supports running the JS generated from Dart in JavaScript engines outside browsers. We use this for testing, some teams use it to run Dart applications in node.js too. Some details:

**compiler**
We don't recommend using the DDC compiler for that. Our recommended approach to use dart2js via `dart compile js --server-mode`. The `--server-mode` flag makes browser based libraries like `dart:html` inaccessible and allows you to use the output of dart2js more easily in those environments. Technically the flag is not required, but it's good practice.

**preambles**
Code generated by the compiler requires a couple APIs to be available from the engine (imagine APIs like `setTimeout`, critical to schedule microtasks and other low-level features in Dart). Some of these APIs are usually found in browsers, but may need to be polyfilled for standalone engines.

For that purpose we require a "preamble" file that is customized for the engine that you will be using. We already provide preambles in the Dart SDK for two common engines [js-shell](https://github.com/dart-lang/sdk/blob/607b3f5ddba34348f965cce99e4cac5e87a46a99/sdk/lib/_internal/js_runtime/lib/preambles/jsshell.js) and [d8](https://github.com/dart-lang/sdk/blob/607b3f5ddba34348f965cce99e4cac5e87a46a99/sdk/lib/_internal/js_runtime/lib/preambles/d8.js). The community currently maintains a preamble for node.js in [`package:node_preamble`](https://pub.dev/packages/node_preamble).

**examples**
[sass](https://github.com/sass/dart-sass) is written in Dart but runs in node.js and is distributed as a npm package. It is a great example that uses `node_preamble`. 

Our team wrote the [generator of `package:web`](https://github.com/dart-lang/web/blob/ff7185c6c0360987116115039a08d9134c8d62aa/tool/generator/README.md) to run in node. This program needed very little from the Dart language, so we ended up creating a custom and over-simplified [preamble](https://github.com/dart-lang/web/blob/ff7185c6c0360987116115039a08d9134c8d62aa/tool/generator/main.mjs) for it. It may not be the best example of a preamble, but it illustrates how it all gets integrated together.

**Comment by maks** [comment]
Thanks so much for all that information @sigmundch ! 👍🏻  to say that it is super useful is an understatement! 

I didn't know about `--server-mode`, thats very helpful too. Nor about the preambles, again very useful!

**Comment by rayan-mazouz** [comment]
This would be fantastic and incredibly useful! It could really help cement Dart as a mainstream language by giving it visibility, since not many languages support WebAssembly well yet.

> We are currently dependent on JS for the following components:
> 
> * Event loop
> * Printing
> * OS-level timer features, such as current time
> * Weak maps, weak references, finalizers
> * Stack traces
> * Conversions between double and String
> * Regexps
> * Math functions (sin, pow, exp, etc.)
> 
> For each of these, we'll have to either:
> 
> * (1) implement the feature in Dart (or intrinsically in the compiler),
> * (2) link to an implementation compiled to Wasm, or
> * (3) import an external implementation from the host.

I think the third option (importing from the host) is necessary and should be implemented first. Just having Dart work without those built-in functionalities would still be extremely valuable ; they may make less sense for non-JS runtimes anyway, and they can be re-implemented by the user using the WASI component model.

For features like math or regex, the first/second approaches make a lot of sense, but it adds complexity, and could be done at a second stage.

IMO, the third approach should come first since it's relatively easy if we just disallow those functions initially.

**Comment by simolus3** [comment]
> For each of these, we'll have to either:

I think option 3 (importing an external definition from the host) would give users the most flexibility.

-  Weak references are impossible to express in WebAssembly alone, so they'll need a (Dart-specific / non-standard) host import anyway.
- Some things like `print` or event loops could be implemented ontop of say WASI definitions, but that then excludes everyone trying to use another runtime or use `dart2wasm` outputs in a modular library.
- Structures like strings and regular expressions could reasonably be implemented in Dart using WASM arrays. But using a host import allows
  - linking (e.g. via `wasm-merge`) an external implementation, saving complexity in the Dart SDK and giving users more options.
  - still support common runtimes like WASI by linking an adapter implementing Dart externs over WASI imports.
  - potentially better performance if string / regex implementations are implemented natively on the host instead of in Dart.
  - re-using the implementation when linking multiple independent programs (potentially even written in different languages).

I have experimented with an approach following option 3, and have a working prototype for this. Tooling for this is in [this repository](https://github.com/simolus3/dart-wasm-tools/) and this [SDK fork](https://github.com/simolus3/dart-sdk/tree/dart2wasm-standalone). For all of the things where JS is currently used, this adds [WASM imports](https://github.com/simolus3/dart-sdk/blob/dart2wasm-standalone/sdk/lib/_internal/wasm_standalone/lib/runtime.dart) instead. 

To play around with this, I've implemented a small runtime using wasmtime that loads modules compiled with `dart2wasm` and injects all the necessary externs. It implements an event loop and timers, printing, current time, stack traces, and some string functions. Enough to get a hello world from Dart running through wasmtime from a native Rust program :)

I think it would be really nice if Dart could eventually support something like this, and I would be happy to start contributing some of these changes. The state in the fork is very messy since I just copied stuff around until it worked, but I can structure this and open smaller CLs towards this. Would contributions adding a `dart2wasm_standalone_platform.dill` using host imports be welcome? If not, or if there's uncertainty about how the SDK should support this, is there a way we as a community can help here to test and evaluate different approaches?

**Comment by mkustermann** [comment]
> I think it would be really nice if Dart could eventually support something like this, and I would be happy to start contributing some of these changes. ...
> Would contributions adding a `dart2wasm_standalone_platform.dill` using host imports be welcome? If not, or if there's uncertainty about how the SDK should support this, is there a way we as a community can help here to test and evaluate different approaches?

As you point out, running dart2wasm code in a non-JS environment will require different implementation of the `dart:*` libraries. To ensure such a different implementation of `dart:*` libraries is correctly done it needs to be tested. It would make sense to run our existing extensive language and core library test suites from the dart-lang/sdk repo on such an implementation. 

It may also require changes to the dart2wasm compiler (which e.g. today is aware of JS environment (it emits mjs files) and may need to be aware of a pure wasm environment as well).

So given this, it would make sense for that to live in dart-lang/sdk and be build & tested using our the infrastructure.

To test the code, it would then also require a non-JS wasm runtime that loads the compiled modules & runs them - the code for which doesn't necessarily need to be in the dart-lang/sdk repo - it could be pulled in via `DEPS` (as we do with other tools e.g.  `binaryen`) - but it would need to have compatible license.

As this would be highly experimental, we wouldn't ship/distribute any of it while it's being developed. If it reaches a high level of maturity and we think we can commit to this & own it, that may change.

We are interested in the idea of running dart2wasm code in non-JS environments but it's not a priority for us right now. Though we could help with some technical guidance, reviewing, etc on the implementation.

**Comment by iapicca** [comment]
@simolus3 any chance that this could be used as stepping stone for [wasi support](https://github.com/dart-lang/sdk/issues/56366)?

**Comment by kevmoo** [comment]
RE @mkustermann 

Just brainstorming:

* Could we ponder a "pure wasm" branch of the `dart:*` libraries where we start implementing things in pure wasm that are currently using JS? As a first step, we (maybe @simolus3) could document all of the places where we rely on JS bits. It might be worth checking that into the REPO and used to track work. Having something setup to allow incremental "chipping away" would be amazing.
* In such a model, we *could* hold off doing a non-JS wasm runtime in the short/med-term. We could just use V8 but incrementally with less JS glue code.

I'll bring this up in the dart2wasm sync we are having today and update here in a few hours.

RE @iapicca 

Yes, this is required for us to move towards Component Model, but the ecosystem is blocked by https://github.com/WebAssembly/component-model/issues/525 – which hasn't seen much progress, sadly.

**Comment by simolus3** [comment]
> To test the code, it would then also require a non-JS wasm runtime that loads the compiled modules & runs them

Technically that "non-JS wasm runtime" could also be a JavaScript program manually instantiating the compiled module 🙃 The idea is that users can write their own runtime/host and Dart apps use that through WebAssembly imports, for tests we can provide those imports from JavaScript without compromising that goal.

> As a first step, we (maybe @simolus3) could document all of the places where we rely on JS bits

Modulo regular expressions which I haven't looked at yet, I think [this](https://github.com/simolus3/dart-sdk/blob/334c921c9f159ee5cc263455278d374217919b19/sdk/lib/_internal/wasm_standalone/lib/runtime.dart#L27-L240) is a complete list.

> @simolus3 any chance that this could be used as stepping stone for https://github.com/dart-lang/sdk/issues/56366?

Yes, I have looked at the component model and WASI, and I think following approach (3) is the shortest path for that.

Basically, the idea is that `dart2wasm` emits function imports for everything Dart needs. These imports don't follow the component model spec and are entirely tailored towards Dart (super unstable, perhaps documented only in source or in a markdown file in the SDK repo somewhere).
Then _some tool_ (and I think it would make sense for that to exist outside of the SDK) could take these wasm modules and turn them into a WASI component.
As an example, an import generated by `dart2wasm` might be `(import "dartruntime" "scheduleAfterMicros" (func (i64 (ref func) anyref) -> (externref)))` and the unstable/informal ABI is "call the provided funcref with the given anyref in an event loop iteration after the duration in microseconds given in the first parameter, return a handle to that timer". I can provide an import for that as a host loading the module, but I can also implement that function using WASI clocks and essentially component-ify the dart2wasm output. I wouldn't even have to wait for the component model to support wasm GC, and the Dart SDK itself can be completely unaware of the component model.

**Comment by mkustermann** [comment]
> Technically that "non-JS wasm runtime" could also be a JavaScript program manually instantiating the compiled module 🙃 The idea is that users can write their own runtime/host and Dart apps use that through WebAssembly imports, for tests we can provide those imports from JavaScript without compromising that goal.

True. Though we need something to test that the `dart:*` libraries are correct.
And any 3rd party embedder would also want to test that their implementation of the wasm imports is correct.

> These imports ...  are entirely tailored towards Dart (super unstable, perhaps documented only in source or in a markdown file in the SDK repo somewhere).

For the JS environment we have a stable mjs API that users can use and it allows the dart2wasm compiler to freely move logic from wasm to mjs or the other way around. It doesn't have a stable interface for the wasm imports it emits in the module.

For the non-JS environment: If the imports aren't stable, then the output of dart2wasm and the embedder are tightly coupled. So that would mean that we wouldn't want our tooling (e.g. `dart compile wasm`) to emit such modules. 

So we'd probably have a different Wasm SDK that uses our existing building blocks but provide it's own CLI for build/test/run, it would pinpoint the Dart SDK commit it uses and then have the corresponding runtime + embedder that provides implementations for the wasm imports.

**Comment by simolus3** [comment]
> For the non-JS environment: If the imports aren't stable, then the output of dart2wasm and the embedder are tightly coupled. So that would mean that we wouldn't want our tooling (e.g. `dart compile wasm`) to emit such modules.

I suppose we can stabilize the imports once we have a complete SDK running on WebAssembly alone. The SDK rarely adds more platform dependencies (weak references are the most recent addition I can remember), so they would probably be stable in practice.
Building a wrapper component to make a Dart module requiring those imports work in say a WASI context could be part of the SDK, but it could also be developed as a third-party module.

> Having something setup to allow incremental "chipping away" would be amazing.

So the approach would be something like this then?

1. Add a `wasm_standalone_platform.dill` that is identical to the normal platform, add a flag to `dart2wasm` to use that.
2. Add separate tests instantiating that module through the emitted `.mjs` file.
3. Incrementally migrate `@JS` declarations to `@pragma('wasm:import')`s, provide definitions in the test runner.
4. Once everything is migrated, disable JS interop on that target and stop emitting `.mjs` files. We can still test everything with a direct `WebAssembly.instantiate` call providing all definitions.
5. Done?

If the approach sounds reasonable to the Dart team, I would be happy to help.

**Comment by mkustermann** [comment]
Apologies for the late reply (this message got buried in my inbox ...)

> So the approach would be something like this then?

Something like that.

Though I'm not sure that we'd want to migrate all core library code that currently uses `@JS()` to `@pragma('wasm:interop')`

If doing so regresses mjs/wasm size, performance or maintainability of the code, then we wouldn't want to do it that way.

Since we'd build a different platform file, we have the flexibility to provide different implementations for core library features, (e.g. different [weak_patch.dart](https://github.com/dart-lang/sdk/blob/7071a5c3e8c9c8174b674ad03c452404fb72e7ae/sdk/lib/_internal/wasm/lib/weak_patch.dart#L62), [math_patch.dart](https://github.com/dart-lang/sdk/blob/7071a5c3e8c9c8174b674ad03c452404fb72e7ae/sdk/lib/_internal/wasm/lib/math_patch.dart#L264), ...)

Also one could have more efficient implementations of some core library features in non-JS environments (e.g. calling to wasm embedder for utf8 decoding instead of shipping our own utf8 decoder, ...).

**Comment by simolus3** [comment]
To give an update on this: As of 58f9d88fb2ee59481364aa86c2693795a5d46e16, it's possible to compile Dart apps to WebAssembly without any JS interop! However, that doesn't mean that you can just `wasmtime run dart_program.wasm` now, we still need external definitions for things like stack traces, regular expressions and [all the other things](https://github.com/dart-lang/sdk/blob/main/sdk/lib/_internal/wasm_standalone/lib/embedder.dart).

Wasmtime has a Rust API that can be used to provide Rust-based implementations of these functions. I'll look at that as an example soon, @kevmoo mentioned he was able to get this working with LLM-generated implementations as well.

Remaining items to get this working with stuff available in released SDKs are:

- https://github.com/dart-lang/sdk/issues/54394 (wasmtime doesn't support legacy try/catch instructions).
- (done!) including `dart2wasm_standalone_platform.dill` in released SDKs, and exposing `--standalone` in `dart compile wasm`.

Then there are also some fairly obvious improvements (which don't block this work though, I can take a look at them):

- Avoid the remaining `js_interop` import (https://github.com/dart-lang/sdk/issues/63166) to remove that library from the standalone platform.
- Evaluate constants in a start function instead of lazily on first access.
- (done!) Restructure patches to group them by whether they apply to all wasm targets or the JS / JS compatibility / standalone target.

Once we have something that works with external host imports, a next step could be to implement some of these imports in WebAssembly itself. This would never support stack traces or weak maps properly because that's fundamentally unavailable, but we could link an existing implementation of regular expressions, for instance. This would give us self-contained WebAssembly modules that _can_ be run with `wasmtime` and other runtimes.

It would also be helpful to hear from those interested in this: Would a Rust crate to help run Dart programs compiled to WebAssembly be useful to you, or do you need something self-contained for an existing runtime that doesn't allow custom host imports (in that case, what runtime are you using?).

**Comment by kevmoo** [comment]
@simolus3 – I copied the list bits of your comment up to the main issue 👆. Makes it easier to track!

**Comment by justinfagnani** [comment]
@simolus3 (and hi @kevmoo! 👋 ) I'm working on a language that has many of the same requirements as Dart for running in Wasm. So far I've taken the approach of having compiler flags for the difference environments, which include one where I've added custom imports for missing WASI features like backtraces and process execution, and in the future workers and possibly weakmaps. There are WASI proposals for some of these things, but it's really unclear how fast they're moving, so it was easier to just implement them in Rust for integration with Wasmtime.

It might be more coordination overhead than you want, but I wanted to put it out there that I'd be interested in collaborating on a Rust crate that provides these types of imports. There's nothing much language-specific about them, though backtraces do require a protocol for handling strings. A well-defined spec for the imports would allow other to implement them for other runtimes as well, and maybe help spur some progress on standardized versions.

**Comment by bivens-dev** [comment]
@justinfagnani just for some additional context if you hadn’t already seen it but there is this repo from @simolus3 where if I understand correctly some of this might already be happening https://github.com/simolus3/wasm.dart

**Comment by simolus3** [comment]
> So far I've taken the approach of having compiler flags for the difference environments, which include one where I've added custom imports for missing WASI features like backtraces and process execution, and in the future workers and possibly weakmaps. There are WASI proposals for some of these things, but it's really unclear how fast they're moving, so it was easier to just implement them in Rust for integration with Wasmtime.

WebAssembly modules are very easy to compose, so even if your imports don't match the ones used by Dart, it would be possible implement a translation layer in WebAssembly and merge that with `dart2wasm` output to turn a module with Dart imports into a module with your imports. This is the approach I use for my WASI experiments too.

The [actual imports](https://github.com/dart-lang/sdk/blob/main/sdk/lib/_internal/wasm/standalone/embedder.dart) are basically a direct port of what Dart imports from JavaScript for the default WebAssembly build. I don't think it makes sense for them to follow any particular spec, but if you have an existing crate I could take a look at what it would take to target that from Dart.
